### PR TITLE
Add: fsdp accelerate version warning

### PR DESCRIPTION
**Pull request by jp1924**
# What does this PR do?

When there was a problem that the model checkpoint could not be saved     
while running fsdp to the latest version of transformers.      
It turned out that it was a problem caused by the low accelerate version.     
I think I should at least warn you about this so that problems like mine don't happen.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@muellerzr
@pacman100

### CODE REVIEW

**Comment by muellerzr** [comment]
cc @pacman100

**Review by pacman100 — COMMENTED** [review]
Hello, can you explain the issue with a minimal reproducer for this. Thank you!

**Comment by jp1924** [comment]
# Reproduction Code
```python
from transformers import (
    TrainingArguments,
    Trainer,
    BertForSequenceClassification,
    BertTokenizer,
)
from datasets import load_dataset
from transformers.utils import check_min_version
from accelerate import __version__ as accelerate_version
from packaging import version

if version.parse(accelerate_version) > version.parse("0.24.1"):
    raise RuntimeError(
        "This error occurs in accelerate versions 0.24.1 and earlier. Please lower your accelerate version"
    )

# This error occurs in transformers 4.36.0 and later
check_min_version("4.36.0")


def main(training_args: TrainingArguments) -> None:
    model_name_or_path = "brenomatos/bert-base-uncased"
    tokenizer = BertTokenizer.from_pretrained(model_name_or_path)
    model = BertForSequenceClassification.from_pretrained(model_name_or_path)

    trainer = Trainer(
        model=model,
        tokenizer=tokenizer,
        args=training_args,
    )
    trainer.save_model()
    trainer.save_state()


if "__main__" in __name__:
    training_args = TrainingArguments(
        output_dir="./output_dir",
        run_name="FSDP_save_test",
        fsdp="full_shard auto_wrap",
        fsdp_transformer_layer_cls_to_wrap="BertEncoder",
    )
    main(training_args)
```
run script
>torchrun --nproc_per_node=4 ./fsdp_error_test.py
# Version
torch                    2.0.1+cu118
transformers             4.36.2
accelerate               0.24.0
# Descripion
First of all, this is an issue I discovered while trying to further pretrain the LLAMA2 13B model.
My computer was running low on memory, so if I tried to save a checkpoint in the middle of training, I would get a CPU OOM and the training would end.
However, saving the checkpoint after the training was over did not cause the OOM, so I had been using `trainer.save_model()` to save the model after all the training was finished. 
However, when I saved the model after updating transformers to the latest version, the model was not saved properly.

I looked up the problem and realized that it was caused by the `version.parse(accelerate_version) > version.parse("0.24.1")` code added in transformers 4.36.0.
I can understand why they added it, but I thought they should have at least included a warning, so I issued a PR.

**Comment by amyeroberts** [comment]
Gentle ping @pacman100

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -2831,6 +2831,12 @@ def save_model(self, output_dir: Optional[str] = None, _internal_call: bool = Fa
                 state_dict = self.accelerator.get_state_dict(self.model)
                 if self.args.should_save:
                     self._save(output_dir, state_dict=state_dict)
+            else:
+                if not version.parse(accelerate_version) > version.parse("0.24.1"):
+                    logger.warning(
+                        "I can't save FSDP normally because the version of accelerate is low!"
+                        " Please increase the version of accelerate to 0.24.5 or higher!"
+                    )
```
Comment (pacman100) [inline src/transformers/trainer.py:2839]:
```suggestion
            elif version.parse(accelerate_version) <= version.parse("0.24.1") and ("FULL_STATE_DICT" in str(self.accelerator.state.fsdp_plugin.state_dict_type)):
                    raise ValueError(
                        "To save full state dict in transformers format, the minimum version of 🤗 Accelerate required is 0.24.2 or higher. Current version: {version.parse(accelerate_version)}."
                    )
```

**Review by pacman100 — COMMENTED** [review]
Thank you @jp1924 for the fix and a reproducer! Please refer to the suggestion wherein we check exactly the case wherein Accelerate version needs to be updated and raise an explicit error when the version is lower.

File: src/transformers/trainer.py (hunk 1 above)
Comment (jp1924) [reply src/transformers/trainer.py:2839]: @pacman100 Thanks for taking me up on my suggestion, it looks good to me! 
I think we should add a relevant warning in training_args.py or trainer.py to let people know about this before they start training, what do you think?

**Pull request closed without merging**